### PR TITLE
Add official signed build pipeline

### PR DESCRIPTION
**Pull request by dagood**
Adds `runtime-official.yml`.

The new pipeline builds the subsets in release mode. It includes an `AllConfigurations` libraries job, which fans out into each installer job in addition to the platform-specific artifacts. After the installer build, a utility stage runs to sign and prepare the artifacts for publish. Then Arcade validation and publish stages run.

I tweaked Libraries publish a bit to let AllConfigurations bits get published. `testhost` doesn't seem to get built that way, causing a copy step to fail, plus publishing in an additional layout to get the OOB packages signed.

Feedback items on the runtime live CI build (https://github.com/dotnet/runtime/issues/749) are not generally addressed in this PR. I changed skiptests from a parameter to a variable so that I could skip Installer tests, but not much else.

I've run this all the way through green, here: https://dev.azure.com/dnceng/internal/_build/results?buildId=460871

~That build didn't have any channels auto-assigned to the branch, so no publishing happened. I'm working on trying out a validation channel in a new build now.~

Tested publishing via the validation channel here: https://dev.azure.com/dnceng/internal/_build/results?buildId=461236. Looks reasonable to me. It didn't publish to `dotnetcli` blob storage but I believe Arcade will do that based on the channel the build's assigned to. `master` is already set up to default-assign to `.NET Core 5 Dev`, so building after merging will try that out.

@dotnet/runtime-infrastructure @mmitche

### CODE REVIEW

File: eng/pipelines/runtime-official.yml [hunk 1]
```diff
@@ -0,0 +1,98 @@
+trigger:
```
Comment (safern) [inline eng/pipelines/runtime-official.yml]: Do we really need a dedicated `runtime-official.yml`? Should we instead use `runtime.yml` and condition configurations and matrix based on `isOfficialBuild`? 

Usually maintaining multiple .yml files that define different matrixes is kind of tricky because a configuration can be forgotten in some yml, such as triggers and stuff like that.

File: eng/pipelines/runtime-official.yml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  batch: true
+  branches:
+    include:
+    - master
+    - release/*
```
Comment (safern) [inline eng/pipelines/runtime-official.yml]: Should we exclude paths? Like .md files and docs/*? I don't think we should trigger an official build for documentation changes.

File: eng/pipelines/runtime-official.yml [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+pr: none
+
+variables:
+- template: /eng/pipelines/common/variables.yml
+# TODO: (Consolidation) Switch away from old signing/validation variables from former Core-Setup.
+- name: TeamName
+  value: dotnet-core-acquisition
```
Comment (safern) [inline eng/pipelines/runtime-official.yml]: Why is that the TeamName? This should just be .NET Core, no? Or .NET Runtime?

File: eng/pipelines/runtime-official.yml (hunk 1 above)
Comment (dagood) [reply eng/pipelines/runtime-official.yml]: > Should we instead use `runtime.yml` and condition configurations and matrix based on `isOfficialBuild`?

I think we should. It will take some extra time to condition the configurations and list of platforms correctly, I don't think we need to do it with the initial implementation, but I'm open to it if you think it's critical.

File: eng/pipelines/official/stages/publish.yml [hunk 4]
```diff
@@ -32,7 +35,7 @@ stages:
     symbolPublishingAdditionalParameters: "'-warnAsError:$false'"
     # Enable SDL validation, passing through values from the 'core-setup-sdl-validation' group.
     SDLValidationParameters:
-      enable: true
+      enable: false # TODO: (Consolidation) Decide who owns SDL validation errors and enable.
```
Comment (safern) [inline eng/pipelines/official/stages/publish.yml]: Can we open issues for these TODOs?

File: eng/Signing.props [hunk 5]
```diff
@@ -1,3 +1,123 @@
...
+
+    <!-- Sign api-ms-win-core-xstate-l2-1-0 binary as it is only catalog signed in the current SDK. -->
+    <ItemsToSign
+      Condition="'$(ConfigurationGroup)' == 'Release' and '$(TargetArchitecture)' == 'x86'"
+      Include="$(CoreCLRArtifactsPath)Redist\ucrt\DLLs\$(TargetArchitecture)\api-ms-win-core-xstate-l2-1-0.dll" />
+
+    <!-- Sign libraries. -->
+    <ItemsToSign Include="$(LibrariesNativeArtifactsPath)*.dll" />
+    <ItemsToSign Include="$(LibrariesSharedFrameworkRefArtifactsPath)*.dll" />
+    <!-- Most runtime artifacts will be crossgenned, so sign them post-crossgen. mscorlib isn't. -->
```
Comment (safern) [inline eng/Signing.props]: I guess the crossgen'd libraries artifacts are signed as part of the nupkg signing?

File: eng/Signing.props [hunk 6]
```diff
@@ -1,3 +1,123 @@
...
+    <FileSignInfo Include="mscordaccore.dll" CertificateName="MicrosoftSHA2" />
+
+    <!-- Sign api-ms-win-core-xstate-l2-1-0 binary as it is only catalog signed in the current SDK. -->
+    <ItemsToSign
+      Condition="'$(ConfigurationGroup)' == 'Release' and '$(TargetArchitecture)' == 'x86'"
+      Include="$(CoreCLRArtifactsPath)Redist\ucrt\DLLs\$(TargetArchitecture)\api-ms-win-core-xstate-l2-1-0.dll" />
+
+    <!-- Sign libraries. -->
+    <ItemsToSign Include="$(LibrariesNativeArtifactsPath)*.dll" />
+    <ItemsToSign Include="$(LibrariesSharedFrameworkRefArtifactsPath)*.dll" />
```
Comment (safern) [inline eng/Signing.props]: Why aren't these artifacts, like the ref and native signed as part of the nupkg? Because they are part of the installers?

File: eng/pipelines/runtime-official.yml (hunk 2 above)
Comment (dagood) [reply eng/pipelines/runtime-official.yml]: Yeah, I guess I can copy the filters over for now.

File: eng/liveBuilds.targets [hunk 7]
```diff
@@ -5,12 +5,29 @@
...
+
+    <CoreCLRSharedFrameworkDir>$([MSBuild]::NormalizeDirectory('$(CoreCLRArtifactsPath)', 'sharedFramework'))</CoreCLRSharedFrameworkDir>
+
+    <LibrariesArtifactsPath Condition="'$(LibrariesArtifactsPath)' == ''">$([MSBuild]::NormalizeDirectory('$(RepoRoot)', 'artifacts'))</LibrariesArtifactsPath>
+
+    <LibrariesPackagesDir>$([MSBuild]::NormalizeDirectory('$(LibrariesArtifactsPath)', 'packages', '$(LibrariesConfiguration)'))</LibrariesPackagesDir>
+    <LibrariesShippingPackagesDir>$([MSBuild]::NormalizeDirectory('$(LibrariesPackagesDir)', 'Shipping'))</LibrariesShippingPackagesDir>
+    <LibrariesNonShippingPackagesDir>$([MSBuild]::NormalizeDirectory('$(LibrariesPackagesDir)', 'NonShipping'))</LibrariesNonShippingPackagesDir>
+
+    <LibrariesAllConfigurationsArtifactsPath Condition="'$(LibrariesAllConfigurationsArtifactsPath)' == ''">$([MSBuild]::NormalizeDirectory('$(RepoRoot)', 'artifacts'))</LibrariesAllConfigurationsArtifactsPath>
```
Comment (safern) [inline eng/liveBuilds.targets]: Why is this needed? These artifacts will end-up as packages in the `Shipping`/`NonShipping`, so I think we could just use `LibrariesShipping/NonShippingPackagesDir` properties.

File: eng/pipelines/runtime-official.yml (hunk 3 above)
Comment (dagood) [reply eng/pipelines/runtime-official.yml]: Yep, this is what the TODO is about, `TODO: (Consolidation) Switch away from old signing/validation variables from former Core-Setup.`

File: eng/pipelines/common/upload-unsigned-artifacts-step.yml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+- task: CopyFiles@2
+  displayName: Prepare job-specific intermediate unsigned artifacts subdirectory
+  inputs:
+    SourceFolder: '$(Build.SourcesDirectory)/artifacts/packages/$(_BuildConfig)'
+    Contents: |
+      Shipping/**/*
+      NonShipping/**/*
+    TargetFolder: '$(Build.StagingDirectory)/UnsignedArtifacts/${{ parameters.name }}'
+    CleanTargetFolder: true
+  condition: and(succeeded(), eq(variables._BuildConfig, 'Release'))
```
Comment (safern) [inline eng/pipelines/common/upload-unsigned-artifacts-step.yml]: Is the _BuildConfig condition in these two jobs necessary? We just import this template when isOfficialBuild == true.

File: eng/Signing.props (hunk 5 above)
Comment (dagood) [reply eng/Signing.props]: No, they're signed in `SignR2RBinaries`. They have to be signed before getting put in the MSIs because Arcade can't recursively sign them (or the bundle installer EXEs).

File: eng/pipelines/runtime-official.yml (hunk 1 above)
Comment (safern) [reply eng/pipelines/runtime-official.yml]: Sounds good. I think once we have the single pipeline landed we can merge these two and try to do it as clean as possible, but having these 2 separated now makes sense, specially because of all the changes it will suffer in the coming days.

File: eng/Signing.props (hunk 6 above)
Comment (dagood) [reply eng/Signing.props]: Yep.

File: eng/pipelines/runtime-official.yml (hunk 3 above)
Comment (safern) [reply eng/pipelines/runtime-official.yml]: Ok, cool. As long all these are tracked by an issue.

File: eng/pipelines/common/upload-unsigned-artifacts-step.yml (hunk 8 above)
Comment (dagood) [reply eng/pipelines/common/upload-unsigned-artifacts-step.yml]: Good call, I believe it's just a carryover from core-setup where Release and Debug were always built--the official build served as rolling test CI as well as the official build. Will remove.

File: eng/Signing.props (hunk 5 above)
Comment (safern) [reply eng/Signing.props]: Ok, cool. The way we need to sign is confusing now, I was used to the signing we had, that we just signed nupkgs 😄

File: eng/pipelines/runtime-official.yml (hunk 1 above)
Comment (dagood) [reply eng/pipelines/runtime-official.yml]: Cool, added this to the list in https://github.com/dotnet/runtime/issues/749 to keep track of it.

File: eng/pipelines/common/upload-unsigned-artifacts-step.yml (hunk 8 above)
Comment (safern) [reply eng/pipelines/common/upload-unsigned-artifacts-step.yml]: Then we can remove the condition as the default when no condition is provided is `succeeded()`.

File: eng/Signing.props (hunk 5 above)
Comment (dagood) [reply eng/Signing.props]: > Ok, cool. The way we need to sign is confusing now

Very. Welcome to the life of core-setup. 😉 At least I didn't need to bring the confusing publishing over.

File: eng/liveBuilds.targets (hunk 7 above)
Comment (dagood) [reply eng/liveBuilds.targets]: > These artifacts will end-up as packages in the `Shipping`/`NonShipping`

This happens for the platform-specific artifacts, but the AllConfigurations artifacts have to be downloaded to a transport directory due to overlaps with the platform-specific artifacts. (Specifically, it caused a build failure with the `docs` directory when I tried to put them both in the same place.)

This property is the entry point for the build to find the transport directory:

```yaml
  # Download allconfigurations bits to a separate location. It overlaps with the live libraries
  # download and causes conflict when unzipping to the same location.
  - ${{ if eq(parameters.useOfficialAllConfigurations, true) }}:
    - name: LibrariesDownloadPathAllConfigurations
      value: 'artifacts/transport/librariesallconfigurations'
    - name: LibrariesAllConfigurationsArtifactsPathArg
      value: /p:LibrariesAllConfigurationsArtifactsPath=${{ parameters.buildCommandSourcesDirectory }}$(LibrariesDownloadPathAllConfigurations)
```

The default is so things will (in theory) work if someone runs both the allconfigurations and platform-specific Libraries builds locally.

Comment (safern) [reply eng/liveBuilds.targets]: I see what you're saying. Since we publish, ref, testhost and other stuff they end-up clashing. 

For official builds we could get away by just publishing /packages/ as part of the artifacts directory and downloading it into `artifacts` in the sources directory of the installer step, however, for local development there is indeed a clash.

What do you think? You think is better having this overhead of more properties that at the end, end-up pointing to the same location by default?

Comment (dagood) [reply eng/liveBuilds.targets]: > For official builds we could get away by just publishing /packages/ as part of the artifacts directory and downloading it into `artifacts` in the sources directory of the installer step,

We can still get clashes that only repro in the official build if, say, the allconfigurations and win-x64 legs start producing a nupkg with the same name. From a defensive perspective I like having the artifacts under completely separate roots for safety and easy traceability in the logs.

> You think is better having this overhead of more properties that at the end, end-up pointing to the same location by default?

IMO the extra properties are not significant vs. other considerations. I think the reason to go for "merging" this is to get local builds as close as possible to AzDO builds. (CoreCLR artifacts should be merged in as well to accomplish this!) But I think this is something that we can do incrementally, and I think it could really use some sort of backstop to make sure artifacts aren't stomping each other without us knowing about it.

Filed https://github.com/dotnet/runtime/issues/1030 for this.

Comment (safern) [reply eng/liveBuilds.targets]: > We can still get clashes that only repro in the official build if, say, the allconfigurations and win-x64 legs start producing a nupkg with the same name. From a defensive perspective I like having the artifacts under completely separate roots for safety and easy traceability in the logs.

I see that very unlikely, and actually the rid specific legs, only produce runtime packages. All configurations when in an official build only produces OOB packages which are not build as part of a rid specific leg. We have been using the same drop for packages for over a year now for libraries builds in (corefx) because that's how we transported all the packages through artifacts to a single job that signed all the nupkgs.

> Filed #1030 for this.

Thanks for filing the issue, I agree that we should get local builds as close as possible to AzDO and have less overhead of passing global properties through command line arguments as that makes things harder to understand and get right.

Comment (dagood) [reply eng/liveBuilds.targets]: > I see that very unlikely, and actually the rid specific legs, only produce runtime packages.

Interesting, yeah, not too familiar with the background here when it comes to the corefx official build. In that case, the backstop is simply the official build failing in that very unlikely case, which is fine. I've added the AllConfigurations part in particular to the https://github.com/dotnet/runtime/issues/749 checklist, no need to flesh out https://github.com/dotnet/runtime/issues/1030 before doing that.

File: eng/Signing.props [hunk 9]
```diff
@@ -1,3 +1,123 @@
...
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(SignBinaries)' == 'true'">
+    <!-- Sign CoreCLR. -->
+    <ItemsToSign Include="$(CoreCLRSharedFrameworkDir)*.dll" />
+    <ItemsToSign Include="$(CoreCLRSharedFrameworkDir)*.exe" />
+
+    <ItemsToSign Include="$(CoreCLRArtifactsPath)System.Private.CoreLib.dll" />
+
+    <ItemsToSign Include="$(CoreCLRArtifactsPath)crossgen2/crossgen2.exe" />
```
Comment (safern) [inline eng/Signing.props]: Do you expect more `.dlls` in these folders? Should we use a wildcard, or are you doing to prevent signing unnecessary files?

This could leak, that if we add a new assembly we could miss signing it and then break the official build and could be kind of hard to diagnose for people not familiar with signing.

File: eng/liveBuilds.targets (hunk 7 above)
Comment (safern) [reply eng/liveBuilds.targets]: Sounds good.

File: eng/pipelines/common/upload-unsigned-artifacts-step.yml [hunk 10]
```diff
@@ -0,0 +1,20 @@
+parameters:
+  name: ''
+
+steps:
+- task: CopyFiles@2
```
Comment (safern) [inline eng/pipelines/common/upload-unsigned-artifacts-step.yml]: Will merging Shipping and NonShipping assets, cause any troubles to loose track which assets are actually Shipping and NonShipping when publishing to BAR?

File: eng/pipelines/libraries/build-job.yml [hunk 11]
```diff
@@ -164,6 +168,14 @@ jobs:
               artifactName: $(librariesBuildArtifactName)
               displayName: Build Assets
 
+          # Save AllConfigurations artifacts using the prepare-signed-artifacts format. The
```
Comment (safern) [inline eng/pipelines/libraries/build-job.yml]: We could avoid publishing all bin directory by conditioning all the steps above when targetFramework eq allConfigurations. Basically adding that condition to the current condition where runTests eq false.

With that we no longer need `_skipTestHostCopy` variable.

File: eng/Signing.props (hunk 9 above)
Comment (dagood) [reply eng/Signing.props]: Yeah, this folder contains a self-contained publish, so there's a lot of stuff here. It's just a straight port from CoreCLR though: https://github.com/dotnet/coreclr/blob/a9f3fc16483eecfc47fb79c362811d870be02249/eng/Signing.props#L9-L18.

File: eng/pipelines/libraries/build-job.yml (hunk 11 above)
Comment (safern) [reply eng/pipelines/libraries/build-job.yml]: I mean, the only needed assets in an official build from an allCOnfigurations leg are the packages, which you're uploading as part of this step.

As follow up of the clean-up, we should be able to remove these, and just upload the packages as part of the artifacts asset.

**Review by safern — APPROVED** [review]
Other than a few comments/questions, LGTM.

File: eng/Signing.props (hunk 9 above)
Comment (safern) [reply eng/Signing.props]: I see. Ok, then we can leave as is.

File: eng/pipelines/common/upload-unsigned-artifacts-step.yml (hunk 10 above)
Comment (dagood) [reply eng/pipelines/common/upload-unsigned-artifacts-step.yml]: > merging Shipping and NonShipping assets

`flattenFolders` defaults to `false`, so this doesn't happen. Would certainly be a problem!

(The rest of the build relies heavily on the implicit `false` too, so I don't think it's worth adding a code comment about it here in particular.)

Comment (safern) [reply eng/pipelines/common/upload-unsigned-artifacts-step.yml]: Ah I missed the `flattenFolders` parameter existed and that it defaulted to false. Cool, thanks for clarifying.

File: eng/pipelines/libraries/build-job.yml (hunk 11 above)
Comment (dagood) [reply eng/pipelines/libraries/build-job.yml]: Good point, I've added a note that this particular publish step goes away when we un-isolate the allconfigurations artifacts.

Comment (safern) [reply eng/pipelines/libraries/build-job.yml]: Thanks for adding the note. Would you mind conditioning all the copy steps above and the upload of those artifacts with a compile time condition expanding the current one to check for framework eq allConfigurations?

Comment (dagood) [reply eng/pipelines/libraries/build-job.yml]: I'd rather leave that until later too... I'm not really familiar with when those copies are used and don't quite follow the change. It seems like incremental improvement to me. This PR is also very close to green after a few hangs, would be nice to get it merged without tinkering with this bit. I'll definitely add it to the note though.

Comment (safern) [reply eng/pipelines/libraries/build-job.yml]: Ok sounds good. We have it as part of the issue so we can cleanup once we get our hands to that issue.

**Comment by safern** [comment]
@dagood I notices that the allConfigurations artifacts are downloaded in every installer build step... is that expected?

Should they only be downloaded in 1 build? 

Which build is the one that will sign them?

**Comment by dagood** [comment]
That's normal, that's what I meant by fan out in this bit of the description:

> It includes an `AllConfigurations` libraries job, which fans out into each installer job in addition to the platform-specific artifacts.

Some of the AllConfigurations packages are used to build the platform-specific installers assets, so all jobs need the bits.

> Which build is the one that will sign them?

All the allconfigurations nupkgs are recursively signed by the "Prepare for Publish" stage. If the installer jobs need to pull out anything and put them in the Windows installers, those specific bits are signed then (because the Windows installers can't be recursively signed later). I don't know off the top of my head what's being pulled in this way though.

There might be some time saving potential there, but AFAIK we don't know yet if unnecessary round trips or redundant signing is worse yet.

**Comment by safern** [comment]
Thanks for explaining. 

> Some of the AllConfigurations packages are used to build the platform-specific installers assets, so all jobs need the bits.

Do we know which ones?

**Comment by dagood** [comment]
> > Some of the AllConfigurations packages are used to build the platform-specific installers assets, so all jobs need the bits.
> 
> Do we know which ones?

@jkoritzinsky might, I think at least `Microsoft.NETCore.Platforms` and `Microsoft.NETCore.Targets`.

**Comment by safern** [comment]
The reason why I ask is because the drop of allConfigurations packages is heavy. So we might instead be able to just upload to a different drop the ones we know are needed.

**Comment by jkoritzinsky** [comment]
The two packages @dagood mentioned are the packages that we need.
